### DISCUSSION TITLE
Adam Optimizer wrong variable type?

### DISCUSSION THREAD

**Discussion by kg492** [Part-II-Computational-Physics/.github#77] [Q&A] [answered]
I will supply my Adam Optimizer for reference: 

def adam(model, n_steps=100):
    for step in range(n_steps):
        
        t = step + 1
    
        # forward
        total_loss, acc = loss2(X, y, model)
        losses.append(total_loss.data)
        accuracies.append(acc)

        # backward
        total_loss.backward()
        
         # update (adam)
        lr = eta * learning_rate(t, n_steps)
        for i, p in enumerate(model.parameters()):
            g = p.grad
            m[i] = b1 * m[i] + (1 - b1) * g
            v[i] = b2 * v[i] + (1 - b2) * g**2
            
            # Bias correction
            m_hat = m[i] / (1 - b1**t)
            v_hat = v[i] / (1 - b2**t)
            
            # Update parameters
            p.data -= p.data - lr * (m_hat / (np.sqrt(v_hat) + eps) + lmbda * p.data)
            
            # Reset gradient
            p.grad = None
            
        if t >= n_steps:
            break

        if step % 10 == 0 or step == n_steps-1:
            print(f"step {step} loss {total_loss.data}, accuracy {acc*100}%")
    
    return losses, accuracies

and then: 

modeladam = initialise(n_neurons = 16)
lossadam, accadam = adam(modeladam)

visualise(modeladam) 
gives me the TypeError: unsupported operand type(s) for +: 'NoneType' and 'float'
It's as if SVM_MMargin_loss() would return None but I printed my loss definition and am getting an output on simple random test:

def loss2(Xdat,ydat,model):
    """Returns the total loss (C(theta) + regularisation) and the accuracy of the neural network."""

    def SVM_MMargin_loss(scores, ydat):
        loss2 = sum([Value.relu(0.1 - yi * scorei) for yi, scorei in zip(ydat, scores)])
        N2 = len([(0.1 - yi*scorei.data) for yi, scorei in zip(ydat, scores)])
        return loss2 * (1/N2)
    
    inputs = [list(map(Value, xrow)) for xrow in Xdat]
    scores = list(map(model, inputs))
 
    data_loss2 = SVM_MMargin_loss(scores, ydat) # or svm_max_margin_loss(scores, ydat)
    
    alpha = 1e-4
    
    reg_loss = alpha * sum((p*p for p in model.parameters()))

    total_loss2 = data_loss2 + reg_loss #loss function + regularisation

    #accuracy = np.mean([yi == np.sign(scorei.data) for yi, scorei in zip(ydat, scores)])
    accuracy = [(yi > 0) == (scorei.data > 0) for yi, scorei in zip(ydat, scores)]
    
    return total_loss2, sum(accuracy) /len(accuracy)

so it should be ok sort of variable? Not certain how to debug this.

**Comment by PatrickC12** [comment, marked as answer]
try placing:

model.zero_grad()

before the line 

 total_loss.backward()

and get rid of:

    # Reset gradient
        p.grad = None

**Reply by kg492** [reply]
well it did get rid of that error so thx!

**Reply by PatrickC12** [reply]
btw couple things with your code:

p.data -= p.data - lr * (m_hat / (np.sqrt(v_hat) + eps) + lmbda * p.data)

consider replacing -= with = here

also

def SVM_MMargin_loss(scores, ydat):
    loss2 = sum([Value.relu(0.1 - yi * scorei) for yi, scorei in zip(ydat, scores)])
    N2 = len([(0.1 - yi*scorei.data) for yi, scorei in zip(ydat, scores)])
    return loss2 * (1/N2)

change to

    def SVM_MMargin_loss(scores, ydat):
        loss2 = [(1- yi * scorei).relu() for yi, scorei in zip(ydat, scores)]
        N2 = len(loss2)
        return sum(loss2) * (1/N2)

think it should work with that
